### ISSUE TITLE
direnv: error .envrc file not found after installing direnv using Snap in zsh

### ISSUE THREAD

**Issue by RobRoseKnows** [direnv/direnv#440]
## Issue description

I installed direnv using snap and although my .envrc file is present, I cannot load it. I am using oh-my-zsh and can successfully run the `direnv` command, but I get the error `direnv: error .envrc file not found` when I try allow or reload even though my .envrc file is definitely there.

### Steps to reproduce

- Install direnv using `snap install direnv`
- Create a .envrc file
- Try to load it.

## Technical details

* `direnv version`: 2.18.2
* OS release: Ubuntu 18.04

**Comment by zimbatm** [comment]
cc @ipolyzos who did the snap packaging

**Comment by ipolyzos** [comment]
@RobRoseKnows  if you please add the following line at the end of the ~/.zshrc file:

> eval "$(direnv hook zsh)"

as per the instructions on https://direnv.net/

**Comment by ipolyzos** [comment]
I have just updated the _https://github.com/ipolyzos/direnv-snap_ repository with the instructions also.

**Comment by ipolyzos** [comment]
@RobRoseKnows could you please help me to replicate your issue, please?

 I can not replicate in my machine or a vm that I have tried...

**Comment by RobRoseKnows** [comment]
@ipolyzos I tried reinstalling direnv and I'm still getting the same error. I'll try in a VM to see if that changes it. This is a fairly new machine though so I don't know what the deal could be. Here's the output of `direnv status` in the meantime:

```
direnv exec path /snap/direnv/12/bin/direnv
DIRENV_CONFIG /home/rob/snap/direnv/12/.config/direnv
DIRENV_BASH /bin/bash
Loaded RC path /home/rob/Dev/projects/fintech/waserverless/.envrc
Loaded RC allowed false
Loaded RC allowPath 
No .envrc found
direnv: error .envrc is blocked. Run `direnv allow` to approve its content. 
```

Does the direnv file need certain ownership or permissions?
I tried a new directory and I'm still getting the same error.
`-rw-r--r-- 1 rob rob 26 Jan 14 16:34 .envrc`

I should also mention I'm getting `direnv: unloading` when I leave that directory.

**Comment by RobRoseKnows** [comment]
Alright I was able to replicate this issue on a fresh DigitalOcean droplet running Ubuntu 18.04. Just follow the default onboarding for a new droplet using Ubuntu 18.04 x64 at the smallest size, wait for the email with the password, login as root and install direnv using snap. Here's everything I did:

```
root@ubuntu-s-1vcpu-1gb-nyc1-01:~# snap install direnv
2019-01-14T21:59:17Z INFO Waiting for restart...
direnv v2.18.2 from 'nullboxorg' installed
root@ubuntu-s-1vcpu-1gb-nyc1-01:~# mkdir test-direnv
root@ubuntu-s-1vcpu-1gb-nyc1-01:~# cd test-direnv/
root@ubuntu-s-1vcpu-1gb-nyc1-01:~/test-direnv# echo export FOO=foo > .envrc
root@ubuntu-s-1vcpu-1gb-nyc1-01:~/test-direnv# cd ..
root@ubuntu-s-1vcpu-1gb-nyc1-01:~# eval "$(direnv hook bash)"
root@ubuntu-s-1vcpu-1gb-nyc1-01:~# cd test-direnv/
direnv: error .envrc is blocked. Run `direnv allow` to approve its content.
root@ubuntu-s-1vcpu-1gb-nyc1-01:~/test-direnv# direnv allow .
direnv: error .envrc file not found
direnv: error .envrc is blocked. Run `direnv allow` to approve its content.
root@ubuntu-s-1vcpu-1gb-nyc1-01:~/test-direnv# direnv status
direnv exec path /snap/direnv/12/bin/direnv
DIRENV_CONFIG /root/snap/direnv/12/.config/direnv
DIRENV_BASH /bin/bash
Loaded RC path /root/test-direnv/.envrc
Loaded RC allowed false
Loaded RC allowPath 
No .envrc found
direnv: error .envrc is blocked. Run `direnv allow` to approve its content.
root@ubuntu-s-1vcpu-1gb-nyc1-01:~/test-direnv# 
```

**Comment by ipolyzos** [comment]
Hi @RobRoseKnows, 

Apologies for my later reply. Indeed seems to be an issue with the snap, even though it works in my machine seems to get issues when i try to install in a new VM i.e.

> $ vagrant init ubuntu/bionic64
> $ vagrant up
> $ vagrant ssh 
> $ snap search direnv
> Name    Version  Publisher   Notes  Summary
> direnv  v2.18.2  nullboxorg  -      direnv -- Unclutter your .profile
> $ sudo snap install direnv
> $ eval "$(direnv hook bash)"
> $ echo "TEST=1" > test/.envrc 
> $ cd test
> direnv: error .envrc is blocked. Run `direnv allow` to approve its content.
> direnv allow .
> direnv: error .envrc file not found
> direnv: error .envrc is blocked. Run `direnv allow` to approve its content.

I will try to dig into this further and will come back soon.

**Comment by ipolyzos** [comment]
Thanks for the advice and sorry for the delay. I have not been able to take a loot on this yet. I will do so at the weekend.

**Comment by ipolyzos** [comment]
Hi @RobRoseKnows , the snap has been fixed using classic confinment. I am working with the snapcraft team to allow it in the officially store but takes some time. Till it get available, you can build and use it using the source at https://github.com/ipolyzos/direnv-snap

**Comment by ipolyzos** [comment]
@RobRoseKnows the snap has been refactored and not officially provided by snapcraft repositories.

feel free to try

**Comment by AntonioFerrete** [comment]
Hello all,

I'm facing the same problem... 
@RobRoseKnows you were able to solve it?
OS release: Ubuntu 19.04

**Comment by IsraelGboluwaga** [comment]
@AntonioFerrete Have you solved yours?

**Comment by IsraelGboluwaga** [comment]
OS: Ubuntu 16.04

```
$ direnv
direnv v2.18.2
Usage: direnv COMMAND [...ARGS]

Available commands
------------------
allow [PATH_TO_RC]:
  Grants direnv to load the given .envrc
deny [PATH_TO_RC]:
  Revokes the authorization of a given .envrc
edit [PATH_TO_RC]:
  Opens PATH_TO_RC or the current .envrc into an $EDITOR and allow
  the file to be loaded afterwards.
exec [DIR] COMMAND [...ARGS]:
  Executes a command after loading the first .envrc found in DIR
help [SHOW_PRIVATE]:
  shows this help
hook SHELL:
  Used to setup the shell hook
prune:
  removes old allowed files
reload:
  triggers an env reload
status:
  prints some debug status information
stdlib:
  Displays the stdlib available in the .envrc execution context
version:
  prints the version
watch [SHELL] PATH:
  Adds a path to the list that direnv watches for changes
direnv: error .envrc is blocked. Run `direnv allow` to approve its content.

$ direnv allow
direnv: error .envrc is blocked. Run `direnv allow` to approve its content.
$ basically anything
basically: command not found
direnv: error .envrc is blocked. Run `direnv allow` to approve its content.
```

I have already added `eval "$(direnv hook bash)"` to the end of my `~/.bashrc` file.

**Comment by IsraelGboluwaga** [comment]
I installed `direnv` with snapd
`sudo snap install direnv` was what I used

**Comment by bedgerotto** [comment]
I installed direnv from snap and, when I run `direnv hook zsh` the response points to: 

> /snap/direnv/54/bin/direnv

But, when I run `which direnv` it points to:

> /var/lib/snapd/snap/bin/direnv

Then, on my ~/.zshrc I written the following, instead of the recomended `eval $(direnv hook zsh)`

```bash
_direnv_hook() {
  # eval "$("/snap/direnv/54/bin/direnv" export zsh)";
  eval "$("/var/lib/snapd/snap/bin/direnv" export zsh)";
}
typeset -ag precmd_functions;
if [[ -z ${precmd_functions[(r)_direnv_hook]} ]]; then
  precmd_functions+=_direnv_hook;
fi
```

And it worked.

**Comment by Kenney-G** [comment]
I am having this issue as well with direnv installed through AUR.

I mean it is correct, there is no .envrc file. There never was. There's a .zshrc file. The plugin never made any ".envrc" file. We were supposed to make one ourselves? If so, how were we to know?

**Comment by ipolyzos** [comment]
The .envrc file is meant to be provided by you indeed. If you please advice https://direnv.net/ for more

**Comment by Splode** [comment]
Navigating to a directory with `.envrc` yielded this error:

```bash
direnv: error .envrc is blocked. Run `direnv allow` to approve its content.
```
This error persists even after running `direnv allow .`

```bash
$ direnv allow .
direnv: error .envrc is blocked. Run `direnv allow` to approve its content.
```

I came across this issue as well when running `direnv` installed from `snap`. Removing the `snap` version and reinstalling with `apt` works as expected.

I'm running Pop!_OS 19.10. `direnv` version from `apt` is `2.20.1`. `direnv` version from `snap` was `2.18.2`

**Comment by IsraelGboluwaga** [comment]
@Splode Run `direnv allow` not `direnv allow .`

**Comment by Splode** [comment]
@IsraelGboluwaga both `direnv allow` and `direnv allow .` yield the same error with the `snap` version. `direnv allow .` with the `apt` version works as expected. Also, the documentations uses `direnv allow .` in the _Quick Demo_.

**Comment by ipolyzos** [comment]
> @IsraelGboluwaga both `direnv allow` and `direnv allow .` yield the same error with the `snap` version. `direnv allow .` with the `apt` version works as expected. Also, the documentations uses `direnv allow .` in the _Quick Demo_.

@Splode which version of the snap you are using? I will try to replicate the behavior.

**Comment by Splode** [comment]
@IsraelGboluwaga 
snap version `2.18.2`
apt version `2.20.1`

**Comment by ipolyzos** [comment]
@Splode @IsraelGboluwaga

have you hooked **direnv** in your shell as per the documentation?

For instruction please refer @ [hook.md|https://github.com/direnv/direnv/blob/master/docs/hook.md]

**Comment by Splode** [comment]
@ipolyzos, yes, I've hooked it into zsh. The only difference in behavior with the 2 scenarios that I can see is the app version and the package distribution (snap vs apt). Using the apt version works as expected with no other changes.

**Comment by IsraelGboluwaga** [comment]
@Splode I noticed that too. Is there a chance something is wrong with the `snap` version?

**Comment by mirao** [comment]
The same issue in bash on Ubuntu 20.04. `direnv allow` works with version installed over `sudo apt install direnv` and doesn't work with snap package.
- snap direnv: v2.18.2
- apt direnv: v2.21.2

**Comment by ipolyzos** [comment]
I will try to replicate this behavior in a new VM

**Comment by anfelor** [comment]
This issue still persists. I see the same behavior as @Splode 
```
snap    2.45.3.1
snapd   2.45.3.1
series  16
ubuntu  20.04
kernel  5.4.0-42-generic
```

**Comment by zimbatm** [comment]
TBH I'm not super convinced of the utility of snap for direnv. It's a single static binary that can be downloaded from the releases: https://github.com/direnv/direnv/releases/ . And because of the nature of direnv, it also is not likely to benefit from the snap sandboxing.

**Comment by RobRoseKnows** [comment]
Yeah I don't use the snap anymore tbh. I just installed it from the releases. I would remove the snap from the README though.

**Comment by LucianU** [comment]
I would remove `direnv` from Snap as well, because the issue still exists in Ubuntu 22.04.

Fortunately, I found this discussion thread quickly, so I didn't spend much time looking for a fix.
